File: qwen-2.5-32B-instruct/quick_sort_Q6_K.cpp
#include <iostream>
#include <vector>

using namespace std;

// Partition function
int partition(vector<int>& arr, int low, int high) {
    // Choose the pivot element (last element in this case)
    int pivot = arr[high];

    // Index of the smaller element
    int i = low - 1;

    // Iterate through the array
    for (int j = low; j < high; j++) {
        // If the current element is smaller than or equal to the pivot
        if (arr[j] <= pivot) {
            // Swap the elements
            i++;
            swap(arr[i], arr[j]);
        }
    }

    // Swap the pivot element with the element at (i + 1)
    swap(arr[i + 1], arr[high]);

    // Return the pivot index
    return i + 1;
}

// Quick sort function
void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        // Partition the array and get the pivot index
        int pivotIndex = partition(arr, low, high);

        // Recursively sort the sub-arrays
        quickSort(arr, low, pivotIndex - 1);
        quickSort(arr, pivotIndex + 1, high);
    }
}


int main() {
    // Sample usage
    vector<int> arr = {5, 2, 9, 1, 7, 6, 3, 8, 4};
    int n = arr.size();

    // Print the original array
    cout << "Original array: ";
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;

    // Sort the array using quick sort
    quickSort(arr, 0, n - 1);

    // Print the sorted array
    cout << "Sorted array: ";
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;

    return 0;
} 

File: qwen-2.5-32B-instruct/quick_sort_Q6_K.py
# Quick-sort implementation in python (4 spaces indentation + detailed comments) and sample usage

# Author: Daniel Luna
# Date: 18-12-2021
# Place: Queretaro, Mexico

# This function is the main one. It takes a list as input and returns the sorted list using quick-sort algorithm.
def quick_sort(list):
    # If the list has 1 or 0 elements, it is already sorted, so we return it
    if len(list) <= 1:
        return list
    # We choose the pivot as the last element of the list
    pivot = list.pop()
    # We create two empty lists, one for the elements less than the pivot and another for the elements greater than the pivot
    less = []
    greater = []
    # We iterate over the list and append the elements to the corresponding list
    for element in list:
        if element < pivot:
            less.append(element)
        else:
            greater.append(element)
    # We return the concatenation of the sorted less list, the pivot and the sorted greater list
    return quick_sort(less) + [pivot] + quick_sort(greater)

# Sample usage
if __name__ == '__main__':
    # We create a list with some elements
    list = [1, 5, 3, 8, 2, 6, 4, 7]
    # We print the list
    print(list)
    # We sort the list and print it
    print(quick_sort(list)) # Output: [1, 2, 3, 4, 5, 6, 7, 8]

# The time complexity of this algorithm is O(n log n) in the average case and O(n^2) in the worst case.
# The space complexity is O(n) because we create two new lists for the elements less than and greater than the pivot.
# The worst case occurs when the pivot is the smallest or the
